pinot-core/src/main/java/org/apache/pinot/core/segment/processing/mapper/SegmentMapper.java

```diff
@@ -95,7 +95,8 @@ public SegmentMapper(List<RecordReaderFileConfig> recordReaderFileConfigs,
         tableConfig.getIndexingConfig().getSortedColumn());
     _fieldSpecs = pair.getLeft();
     _numSortFields = pair.getRight();
-    _includeNullFields = tableConfig.getIndexingConfig().isNullHandlingEnabled();
+    _includeNullFields =
+        schema.isEnableColumnBasedNullHandling() || tableConfig.getIndexingConfig().isNullHandlingEnabled();
```

Comment:
This is one of the two expected changes in semantic. This means that when `SegmentMapper` is used with column based null handling it will keep null values when writing rows. The main issue is that it will treat all columns as nullable.

Reply:
What will happen is that we are going to create a null vector for these not nullable columns. That null vector will be ignored at query time. That is not problematic because it is happening right now. Imagine we have a table that is nullable and then we change it to be not nullable, then already committed segments will have that null vector column but the query engine does just ignore that vector.